core/src/main/java/io/kestra/core/storages/kv/InternalKVStore.java

```diff
@@ -0,0 +1,123 @@
+
+    private static final Logger log = LoggerFactory.getLogger(InternalKVStore.class);
+
+    private final String namespace;
+    private final String tenant;
+    private final StorageInterface storage;
+    private final Logger logger;
+
+    /**
+     * Creates a new {@link InternalKVStore} instance.
+     *
+     * @param namespace The namespace
+     * @param storage   The storage.
+     */
+    public InternalKVStore(final String tenant, final String namespace, final StorageInterface storage) {
+        this(log, tenant, namespace, storage);
+    }
+
+    /**
+     * Creates a new {@link InternalKVStore} instance.
+     *
+     * @param logger    The logger to be used by this class.
+     * @param namespace The namespace
+     * @param tenant    The tenant.
+     * @param storage   The storage.
+     */
+    public InternalKVStore(final Logger logger, @Nullable final String tenant, final String namespace, final StorageInterface storage) {
+        this.logger = Objects.requireNonNull(logger, "logger cannot be null");
+        this.namespace = Objects.requireNonNull(namespace, "namespace cannot be null");
+        this.storage = Objects.requireNonNull(storage, "storage cannot be null");
+        this.tenant = tenant;
+    }
+
+    @Override
+    public String namespace() {
+        return this.namespace;
+    }
+
+    @Override
+    public void putRaw(String key, KVStoreValueWrapper<String> kvStoreValueWrapper) throws IOException {
+        this.validateKey(key);
+
+        this.storage.put(this.tenant, this.storageUri(key), new StorageObject(
+            kvStoreValueWrapper.metadataAsMap(),
+            new ByteArrayInputStream(kvStoreValueWrapper.value().getBytes())
+        ));
+    }
+
+    @Override
+    public Optional<String> getRaw(String key) throws IOException, ResourceExpiredException {
+        this.validateKey(key);
+
+        StorageObject withMetadata;
+        try {
+            withMetadata = this.storage.getWithMetadata(this.tenant, this.storageUri(key));
+        } catch (FileNotFoundException e) {
+            return Optional.empty();
+        }
+        KVStoreValueWrapper<String> kvStoreValueWrapper = KVStoreValueWrapper.from(withMetadata);
+
+        Instant expirationDate = kvStoreValueWrapper.kvMetadata().getExpirationDate();
+        if (expirationDate != null && Instant.now().isAfter(expirationDate)) {
+            this.delete(key);
+
+            throw new ResourceExpiredException("The requested value has expired");
+        }
+        return Optional.of(kvStoreValueWrapper.value());
+    }
+
+    @Override
+    public boolean delete(String key) throws IOException {
+        this.validateKey(key);
+
+        return this.storage.delete(this.tenant, this.storageUri(key));
+    }
+
+    @Override
+    public List<KVEntry> list() throws IOException {
+        List<FileAttributes> list;
+        try {
+            list = this.storage.list(this.tenant, this.storageUri(null));
```

Comment:
Should it filter expired?

Reply:
it does [here](https://github.com/kestra-io/kestra/pull/4217/files/45a399d46046447513e97601998d419ee125310d#diff-68ea1384a14a99e0edabcb694c7538d7b19652c14ad4075dc7d1b20124ff43e9R120)